theano/configdefaults.py

```diff
@@ -42,6 +42,7 @@ def floatX_convert(s):
              "Note: float16 support is experimental, use at your own risk.",
              EnumStr('float64', 'float32', 'float16',
                      convert=floatX_convert,),
+             in_c_key=False
              )
```

Comment:
Now CAReduce don't do that and CAReduceDtype have the output and acc dtype as parameters. I don't see any use of floatX in the file tensor/elemwise.py that would cause this.

If the output dtype of a node depend of floatX or the casting policy that we didn't finish, then I see it can cause problem, as we don't add in the key the output variable.

I find this bad that we need floatX in the in_c_key. This cause too much key when we change floatX for no good reason. I prefer that we add the node outputs type in the key. I made an issue about this: 

https://github.com/Theano/Theano/issues/4466

I just removed that change from this PR.